File: Homework02/sudoku.go
package main

import (
	"fmt"
	"io/ioutil"
	"math/rand"
	"path/filepath"
	"time"
)

func Max(x, y int) int {
	if x < y {
		return y
	}
	return x
}

func Min(x, y int) int {
	if x > y {
		return y
	}
	return x
}
func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}
func readSudoku(filename string) ([][]byte, error) {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	grid := group(filter(data), 9)
	return grid, nil
}

func filter(values []byte) []byte {
	filtered_values := make([]byte, 0)
	for _, v := range values {
		if (v >= '1' && v <= '9') || v == '.' {
			filtered_values = append(filtered_values, v)
		}
	}
	return filtered_values
}

func display(grid [][]byte) {
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid); j++ {
			fmt.Print(string(grid[i][j]))
		}
		fmt.Println()
	}
}

func group(values []byte, n int) [][]byte {

	var additional int = boolToInt(len(values)%n != 0)
	var ans = make([][]byte, len(values)/n+additional)

	var l int
	for l = 0; l < len(ans); l++ {
		ans[l] = make([]byte, n)
	}

	var i, j int
	for i = 0; i < len(ans); i++ {
		for j = 0; j < n; j++ {
			if n*i+j >= len(values) {
				break
			}
			ans[i][j] = values[n*i+j]
		}

	}
	return ans
}

func getRow(grid [][]byte, row int) []byte {

	return grid[row]
}

func getCol(grid [][]byte, col int) []byte {
	var ans = make([]byte, len(grid))
	for i := 0; i < len(grid); i++ {
		ans[i] = grid[i][col]
	}
	return ans
}

func getBlock(grid [][]byte, row int, col int) []byte {
	var br int = row / 3 * 3
	var bc int = col / 3 * 3
	var ans []byte
	ans = make([]byte, 9)
	var counter int = 0
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			ans[counter] = grid[br+i][bc+j]
			counter++
		}
	}
	return ans
}

func findEmptyPosition(grid [][]byte) (int, int) {
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[i]); j++ {
			if grid[i][j] == '.' {
				return i, j
			}
		}
	}
	return -1, -1
}

func contains(values []byte, search byte) bool {
	for _, v := range values {
		if v == search {
			return true
		}
	}
	return false
}

func findPossibleValues(grid [][]byte, row int, col int) []byte {
	var set_of_used = make(map[byte]bool)
	var field []byte = getBlock(grid, row, col)
	for i := 0; i < len(field); i++ {
		set_of_used[field[i]] = true
	}
	var rows []byte = getRow(grid, row)
	for i := 0; i < len(rows); i++ {
		set_of_used[rows[i]] = true
	}
	var cols []byte = getCol(grid, col)
	for i := 0; i < len(cols); i++ {
		set_of_used[cols[i]] = true
	}
	var ok bool
	var i byte
	var counter int = 0
	var possible_vals []byte
	for i = 49; i < 58; i++ {
		_, ok = set_of_used[i]
		if !ok {
			counter++
		}
	}
	possible_vals = make([]byte, counter)
	counter = 0

	for i = 49; i < 58; i++ {
		_, ok = set_of_used[i]
		if !ok {
			possible_vals[counter] = i
			counter++
		}
	}
	return possible_vals
}

func solve(grid [][]byte) ([][]byte, bool) {
	var row, col int
	row, col = findEmptyPosition(grid)
	if row == -1 && col == -1 {
		return grid, true
	}
	var solved bool
	var values []byte = findPossibleValues(grid, row, col)
	for i := 0; i < len(values); i++ {
		grid[row][col] = values[i]
		grid, solved = solve(grid)
		if solved {
			return grid, true
		}
	}
	grid[row][col] = '.'
	return grid, false
}

func checkSolution(grid [][]byte) bool {
	var ok bool
	var c byte
	for i := 0; i < len(grid); i++ {

		var set_of_used = make(map[byte]bool)
		var cols []byte = getCol(grid, i)
		for j := 0; j < len(cols); j++ {
			set_of_used[cols[j]] = true
		}
		for c = 49; c < 58; c++ {
			_, ok = set_of_used[c]
			if !ok {

				return false
			}
		}

	}
	for i := 0; i < len(grid); i++ {
		var set_of_used = make(map[byte]bool)
		var rows []byte = getRow(grid, i)
		for j := 0; j < len(rows); j++ {
			set_of_used[rows[j]] = true
		}
		for c = 49; c < 58; c++ {
			_, ok = set_of_used[c]
			if !ok {

				return false
			}
		}

	}
	var rind = []int{0, 3, 6}
	var cind = []int{0, 3, 6}

	for ii := 0; ii < len(rind); ii++ {
		for jj := 0; jj < len(cind); jj++ {
			var set_of_used = make(map[byte]bool)
			var block []byte = getBlock(grid, rind[ii], cind[jj])
			for j := 0; j < len(block); j++ {
				set_of_used[block[j]] = true
			}
			for c = 49; c < 58; c++ {
				_, ok = set_of_used[c]
				if !ok {
					return false
				}
			}

		}

	}
	return true
}

func generateSudoku(N int) [][]byte {

	var template [][]byte
	
	for i:= 0; i<9; i++{
		template[i] = make([]byte, 9)
		for j:=0; j<9; j++{
		template[i][j] = '.'
		}
	}
	var grid, _ = solve(template)
	s1 := rand.NewSource(time.Now().UnixNano())
	r1 := rand.New(s1)
	for N = 81 - Min(81, Max(N, 0)); N != 0; N += 0 {
		var row int = r1.Intn(8)
		var col int = r1.Intn(8)
		if grid[row][col] == '.' {
			continue
		} else {
			N--
		}
	}
	return grid
}

func main() {
	puzzles, err := filepath.Glob("puzzle*.txt")
	if err != nil {
		fmt.Printf("Could not find any puzzles")
		return
	}
	for _, fname := range puzzles {
		go func(fname string) {
			grid, _ := readSudoku(fname)
			solution, _ := solve(grid)
			fmt.Println("Solution for", fname)
			display(solution)

			fmt.Println(checkSolution(solution))
		}(fname)
	}
	var input string
	fmt.Scanln(&input)
	/*
	   	var line = []byte{1,2,3,4}
	   	var res [][]byte
	   res=group(line,3)
	   for i:=0; i< len(res); i++ {
	   	for j:=0; j<len(res[i]); j++ {
	   		fmt.Println(res[i][j])
	   	}
	   	fmt.Println("ss")
	   }
	   var test = [][]byte{
	   								{'1', '2', '.'},
	   								{'4', '5', '6'},
	   								{'7', '8', '9'}}
	   	fmt.Println(findPossibleValues(test, 0,2))
	*/
}

File: Homework02/sudoku.py
import random
from typing import Tuple, List, Set, Optional


def read_sudoku(filename: str) -> List[List[str]]:
    """ Прочитать Судоку из указанного файла """
    with open(filename) as f:
        content = f.read()
    digits = [c for c in content if c in '123456789.']
    grid = group(digits, 9)
    return grid


def display(grid: List[List[str]]) -> None:
    """Вывод Судоку """
    width = 2
    line = '+'.join(['-' * (width * 3)] * 3)
    for row in range(9):
        print(''.join(grid[row][col].center(width) + ('|' if str(col) in '25' else '') for col in range(9)))
        if str(row) in '25':
            print(line)
    print()


def group(values: List[str], n: int) -> List[List[str]]:
    """
    Сгруппировать значения values в список, состоящий из списков по n элементов

    >>> group([1,2,3,4], 2)
    [[1, 2], [3, 4]]
    >>> group([1,2,3,4,5,6,7,8,9], 3)
    [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    """

    ans = []
    subgroup = []
    for i in range(1, len(values) + 1):
        subgroup.append(values[i - 1])
        if i % n == 0:
            ans.append(subgroup)
            subgroup = []
    if not subgroup == []:
        ans.append(subgroup)

    return ans


def get_row(grid: List[List[str]], pos: Tuple[int, int]) -> List[str]:

    """ Возвращает все значения для номера строки, указанной в pos
    >>> get_row([['1', '2', '.'], ['4', '5', '6'], ['7', '8', '9']], (0, 0))
    ['1', '2', '.']
    >>> get_row([['1', '2', '3'], ['4', '.', '6'], ['7', '8', '9']], (1, 0))
    ['4', '.', '6']
    >>> get_row([['1', '2', '3'], ['4', '5', '6'], ['.', '8', '9']], (2, 0))
    ['.', '8', '9']
    """
    row, _ = pos
    return grid[row]


def get_col(grid: List[List[str]], pos: Tuple[int, int]) -> List[str]:
    """ Возвращает все значения для номера столбца, указанного в pos
    >>> get_col([['1', '2', '.'], ['4', '5', '6'], ['7', '8', '9']], (0, 0))
    ['1', '4', '7']
    >>> get_col([['1', '2', '3'], ['4', '.', '6'], ['7', '8', '9']], (0, 1))
    ['2', '.', '8']
    >>> get_col([['1', '2', '3'], ['4', '5', '6'], ['.', '8', '9']], (0, 2))
    ['3', '6', '9']
    """
    _, col = pos
    return [grid[row][col] for row in range(len(grid))]


def get_block(grid: List[List[str]], pos: Tuple[int, int]) -> List[str]:
    """ Возвращает все значения из квадрата, в который попадает позиция pos
    >>> grid = read_sudoku('puzzle1.txt')
    >>> get_block(grid, (0, 1))
    ['5', '3', '.', '6', '.', '.', '.', '9', '8']
    >>> get_block(grid, (4, 7))
    ['.', '.', '3', '.', '.', '1', '.', '.', '6']
    >>> get_block(grid, (8, 8))
    ['2', '8', '.', '.', '.', '5', '.', '7', '9']
    """
    row, col = pos
    b_row = int(row / 3) * 3
    b_col = int(col / 3) * 3
    return [grid[b_row+row][b_col+col] for row in range(3) for col in range(3)]


def find_empty_positions(grid: List[List[str]]) -> Optional[Tuple[int, int]]:
    """ Найти первую свободную позицию в пазле

    >>> find_empty_positions([['1', '2', '.'], ['4', '5', '6'], ['7', '8', '9']])
    (0, 2)
    >>> find_empty_positions([['1', '2', '3'], ['4', '.', '6'], ['7', '8', '9']])
    (1, 1)
    >>> find_empty_positions([['1', '2', '3'], ['4', '5', '6'], ['.', '8', '9']])
    (2, 0)
    """
    n = len(grid)

    for row in range(0, n):
        for col in range(0, n):
            if grid[row][col] == '.':
                return (row, col)
    return (-42, -42)


def find_possible_values(grid: List[List[str]], pos: Tuple[int, int]) -> Set[str]:
    """ Вернуть множество всех возможных значения для указанной позиции
    >>> grid = read_sudoku('puzzle1.txt')
    >>> values = find_possible_values(grid, (0,2))
    >>> set(values) == {'1', '2', '4'}
    True
    >>> values = find_possible_values(grid, (4,7))
    >>> set(values) == {'2', '5', '9'}
    True
    """
    used_values = set()
    row, col = pos
    field = get_block(grid, (row, col))
    for line in field:
        for value in line:
            used_values.update(value)
    used_values.update(get_row(grid, (row, col)))
    used_values.update(get_col(grid, (row, col)))
    return [value for value in '123456789' if value not in used_values]


def solve(grid: List[List[str]]) -> Optional[List[List[str]]]:
    """ Решение пазла, заданного в grid
    Как решать Судzоку?

    1. Найти свободную позицию
    2. Найти все возможные значения, которые могут находиться на этой позиции
    3. Для каждого возможного значения:
        3.1. Поместить это значение на эту позицию
        3.2. Продолжить решать оставшуюся часть пазла
    >>> grid = read_sudoku('puzzle1.txt')
    >>> solve(grid)
    [['5', '3', '4', '6', '7', '8', '9', '1', '2'], ['6', '7', '2', '1', '9', '5', '3', '4', '8'], ['1', '9', '8', '3', '4', '2', '5', '6', '7'], ['8', '5', '9', '7', '6', '1', '4', '2', '3'], ['4', '2', '6', '8', '5', '3', '7', '9', '1'], ['7', '1', '3', '9', '2', '4', '8', '5', '6'], ['9', '6', '1', '5', '3', '7', '2', '8', '4'], ['2', '8', '7', '4', '1', '9', '6', '3', '5'], ['3', '4', '5', '2', '8', '6', '1', '7', '9']]
    """
    (row, col) = find_empty_positions(grid)
    if((row, col) == (-42, -42)):
        return grid

    values = find_possible_values(grid, (row, col))
    for value in values:
        grid[row][col] = value
        solved = solve(grid)
        if solved:
            return solved

    grid[row][col] = '.'
    return None


def check_solution(solution: List[List[str]]) -> bool:
    for row in range(len(solution)):
        values = set(get_row(solution, (row, 0)))
        if values != set('123456789'):
            return False

    for col in range(len(solution)):
        values = set(get_col(solution, (0, col)))
        if values != set('123456789'):
            return False

    for row in (0, 3, 6):
        for col in (0, 3, 6):
            values = set(get_block(solution, (row, col)))
            if values != set('123456789'):
                return False
    return True


def generate_sudoku(N: int) -> List[List[str]]:
    """ Генерация судоку заполненного на N элементов
    >>> grid = generate_sudoku(40)
    >>> sum(1 for row in grid for e in row if e == '.')
    41
    >>> solution = solve(grid)
    >>> check_solution(solution)
    True
    >>> grid = generate_sudoku(1000)
    >>> sum(1 for row in grid for e in row if e == '.')
    0
    >>> solution = solve(grid)
    >>> check_solution(solution)
    True
    >>> grid = generate_sudoku(0)
    >>> sum(1 for row in grid for e in row if e == '.')
    81
    >>> solution = solve(grid)
    >>> check_solution(solution)
    True
    """
    new_grid = solve([['.'] * 9 for _ in range(9)])
    N = 81 - min(81, max(0, N))
    while N > 0:
        row = random.randint(0, 8)
        col = random.randint(0, 8)
        if new_grid[row][col] != '.':
            new_grid[row][col] = '.'
            N -= 1
    return new_grid


def run_solve(fname):
    grid = read_sudoku(fname)
    start = time.time()
    solve(grid)
    end = time.time()


if __name__ == '__main__':
    grid = read_sudoku('puzzle1.txt')
    display(solve(grid))
